### DISCUSSION TITLE
maximum recursion depth exceeded while calling a Python object

### DISCUSSION THREAD

**Discussion by Karel-Tavernier** [plasma-umass/scalene#592] [Q&A] [answered]
I recently installed scalene on Windows 11, Python 3.11, default install I believe. I tested the following very simple program.
_zonk = 997
_for _ in range(100_000_000):
    zonk = (zonk*23 + 10_007)%10_000
print('Final = ', zonk)_
If I run it directly from the commandline and via scalene I get the following:
__C:\Users\kt\OneDrive\Transfer\Python\intersections\v2023.03.29 without gerber_writer\src>py scalene_test.py
Final =  997
C:\Users\kt\OneDrive\Transfer\Python\intersections\v2023.03.29 without gerber_writer\src>scalene --src scalene_test.py
Error in program being profiled:
 maximum recursion depth exceeded while calling a Python object
Scalene: Program did not run for long enough to profile.
NOTE: The GPU is currently running in a mode that can reduce Scalene's accuracy when reporting GPU utilization.
Run once as Administrator or root (i.e., prefixed with `sudo`) to enable per-process GPU accounting._

Any suggestions how to get this to work?

**Comment by emeryberger** [comment]
I ran your test program successfully on Mac and Windows. Might need some version info to see what's going on. Here's the exact code I used.

```
zonk = 997
for _ in range(100_000_000):
    zonk = (zonk*23 + 10_007)%10_000
print('Final = ', zonk)
```

I used the command line `python3 -m scalene zonk.py`. I'm using version 3.10 on Windows 11, 3.11.1 on Mac, version 1.5.20 of Scalene. Can you try 3.10 on Windows to see if anything changes?

(Minor note, Scalene will ignore the `--src` flag you are passing in (it's not a Scalene option).)

**Comment by emeryberger** [comment]
Just ran successfully on Windows 11 with Python 3.10.10.

**Comment by Karel-Tavernier** [comment]
I tried it on 3.10, and indeed it works fine.
![Screenshot 2023-03-30 230856](https://user-images.githubusercontent.com/68160412/228965630-f3443389-5fc3-4222-8160-f0be04647987.png)

I suppose you are very busy, but may I ask if you intend to make it work on 3.11 in the reasonable short term? I would rather wait a while than return to 3.10 myself.

**Reply by emeryberger** [reply]
Works fine on Windows 11 with Python 3.11.2.

**Comment by Karel-Tavernier** [comment]
Thanks for the quick update.

I upgraded to Python 3.11.2, rebooted for good measure.  No luck.  Now I get an 'Unknown Error'.

> C:\Users\kt\OneDrive\Transfer\Python\intersections\v2023.03.29 without gerber_writer\src>py
> Python 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)] on win32
> Type "help", "copyright", "credits" or "license" for more information.
> C:\Users\kt\OneDrive\Transfer\Python\intersections\v2023.03.29 without gerber_writer\src>py zonk.py
> Final =  997
> C:\Users\kt\OneDrive\Transfer\Python\intersections\v2023.03.29 without gerber_writer\src>py -m scalene zonk.py
> Error in program being profiled:
>  Unknown Error
> Scalene: Program did not run for long enough to profile.
> NOTE: The GPU is currently running in a mode that can reduce Scalene's accuracy when reporting GPU utilization.
> Run once as Administrator or root (i.e., prefixed with `sudo`) to enable per-process GPU accounting.

**Comment by emeryberger** [comment]
I cannot reproduce this error no matter what I do. I'm on ARM64, running Python installed via the installer on python.org. I do not have a GPU on this machine (a Mac). Maybe that is playing a role. Try running with `--cpu`.

**Comment by Karel-Tavernier** [comment]
It is puzzling.  This morning I updated Windows, drivers etc. The error is now again maximum depth. I just tried --src, no difference.

> C:\Users\kt\OneDrive\Transfer\Python\intersections\v2023.03.29 without gerber_writer\src>py -m scalene zonk.py
> Error in program being profiled:
>  maximum recursion depth exceeded
> Scalene: Program did not run for long enough to profile.
> NOTE: The GPU is currently running in a mode that can reduce Scalene's accuracy when reporting GPU utilization.
> Run once as Administrator or root (i.e., prefixed with `sudo`) to enable per-process GPU accounting.
> C:\Users\kt\OneDrive\Transfer\Python\intersections\v2023.03.29 without gerber_writer\src>py -m scalene --src zonk.py
> Error in program being profiled:
>  maximum recursion depth exceeded while calling a Python object
> Scalene: Program did not run for long enough to profile.
> NOTE: The GPU is currently running in a mode that can reduce Scalene's accuracy when reporting GPU utilization.
> Run once as Administrator or root (i.e., prefixed with `sudo`) to enable per-process GPU accounting.

**Comment by Karel-Tavernier** [comment]
Scalene seems to run the intended program. If I give an non-existing file, Scalene complains about it.
> C:\Users\kt\OneDrive\Transfer\Python\intersections\repository\src>py -m scalene --src zap.py
> Scalene: could not find input file zap.py

**Comment by emeryberger** [comment]
`--src` is not an argument that Scalene accepts. Try:

`py -m scalene --cpu zonk.py`

**Reply by Karel-Tavernier** [reply]
Sorry, typo. But it makes no difference:
> C:\Users\kt\OneDrive\Transfer\Python\intersections\v2023.03.29 without gerber_writer\src>py -m scalene --cpu zonk.py
> Error in program being profiled:
>  maximum recursion depth exceeded
However!  I created the following highly sophisticated test program zonk2.py. 

> print('Started')
> zonk = 997
> print('Seed', zonk)
> zonk = (zonk*23 + 10_007)%10_000
> print('First', zonk)
> for _ in range(100_000_000):
>     zonk = (zonk*23 + 10_007)%10_000
> print('Final = ', zonk)
And here is the result. The long for or big range seems to be the problem. (My real problem also has big lists.)
> C:\Users\kt\OneDrive\Transfer\Python\intersections\v2023.03.29 without gerber_writer\src>py -m scalene --cpu zonk2.py
> Started
> Seed 997
> First 2938
> Error in program being profiled:
>  maximum recursion depth exceeded

**Comment by Karel-Tavernier** [comment]
It must have something to do with the range. If I run it with range(300) I get the following.
> C:\Users\kt\OneDrive\Transfer\Python\intersections\v2023.03.29 without gerber_writer\src>py -m scalene --cpu zonk3.py
> Started
> Seed 997
> First 2938
> Final =  4938
> Scalene: Program did not run for long enough to profile.
Scalene seems to run fine, complaining about not running long enough is fair.

**Comment by emeryberger** [comment]
I just tried this on a fresh Windows 11 instance on Azure (x64), Python 3.11.2, works fine. Are you on a dedicated or shared machine? Can you run as admin?

**Reply by Karel-Tavernier** [reply]
I am on a dedicated portable. I ran the cmd as administrator. Same.
Thanks for helping out.

> C:\Users\kt\OneDrive\Transfer\Python\intersections\v2023.03.29 without gerber_writer\src>py -m scalene --cpu zonk2.py
> Started
> Seed 997
> First 2938
> Error in program being profiled:
>  maximum recursion depth exceeded while calling a Python object

**Comment by Karel-Tavernier** [comment]
Another observation. This runs fine.
zap = ''
for _ in range(30_000):
	zap += 'z
If I change it to range(300_000) I again get complaints that the stack depth is exceeded.

**Comment by Karel-Tavernier** [comment]
35_000 works seemingly; 70_000 gives complaints about stack depth.  Something to do with a stack depth setting somewhere?

**Reply by emeryberger** [reply]
Can we move this discussion to the Scalene Slack? Might be more productive. https://join.slack.com/t/scaleneprofil-jge3234/shared_invite/zt-110vzrdck-xJh5d4gHnp5vKXIjYD3Uwg

**Reply by Karel-Tavernier** [reply]
I will try. No guarantees offered.

**Comment by Karel-Tavernier** [comment, marked as answer]
To all. Mr. Berger was very helpful. What happened I initially installed scalene under Python 3.11.0 - problems. Then I upgraded to Python 3.11.2 - still problems. Mr. Berger did a number or experiments on my computer, he suggested to uninstall scalene and reinstall it under 3.11.2, and everything worked fine.
